dev/AppNotifications/AppNotificationTelemetry.h

```diff
@@ -0,0 +1,319 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+#include <WindowsAppRuntimeInsights.h>
+#include <wrl\wrappers\corewrappers.h>
+
+DECLARE_TRACELOGGING_CLASS(AppNotificationTelemetryProvider,
+    "Microsoft.WindowsAppSDK.Notifications.AppNotificationTelemetry",
+    // {1825c850-a487-537d-b768-f0ab298d2565}
+    (0x1825c850, 0xa487, 0x537d, 0xb7, 0x68, 0xf0, 0xab, 0x29, 0x8d, 0x25, 0x65));
+
+
+class AppNotificationTelemetry : public wil::TraceLoggingProvider
+{
+    IMPLEMENT_TELEMETRY_CLASS(AppNotificationTelemetry, AppNotificationTelemetryProvider);
+
+public:
+    DEFINE_EVENT_METHOD(RegisterByAPI)(
+        winrt::hresult hr,
+        std::wstring const& appId) noexcept try
+    {
+        if (c_maxEventLimit >= UpdateLogEventCount())
+        {
+            TraceLoggingClassWriteMeasure(
+                "RegisterByAPI",
+                TelemetryPrivacyDataTag(PDT_ProductAndServicePerformance),
+                _GENERIC_PARTB_FIELDS_ENABLED,
+                TraceLoggingHexUInt32(hr, "OperationResult"),
+                TraceLoggingWideString(appId.c_str(), "AppId"),
+                TraceLoggingBool(IsPackagedApp(), "IsAppPackaged"),
+                TraceLoggingWideString(GetAppName(), "AppName"));
+        }
+    }
+    CATCH_LOG()
+
+    DEFINE_EVENT_METHOD(UnregisterByAPI)(
+        winrt::hresult hr,
+        std::wstring const& appId) noexcept try
+    {
+        if (c_maxEventLimit >= UpdateLogEventCount())
+        {
+            TraceLoggingClassWriteMeasure(
```

Comment:
I remember @brflynn demonstrated that we can get file and line number of failure origin being logged as part of the event. Is this something we can have as part of WinAppSDK telemetry?

Reply:
Do you have an example? We'd have to pass the file name and line number at the call site or wrapped the call into a macro that does it, no?

Reply:
I think wil can do part of the work for you. I'll share with you the PR offline.

Reply:
Thanks for the doc. I'll add that info to the task for adding comprehensive telemetry for toast and push. This is beyond the goal of this PR, which is to provide basic telemetry to measure feature adoption rate.

Reply:
I added @brflynn to see if he has comments on this. He suggested a completely different pattern to use in a recent PR for aggregation.